```shell
sh$ ulimit -n          // show the number
```

当我们打开了句柄（如文件），却忘记了关闭。
```
func main() {
    fmt.Println("PID: ", os.Getpid())
    f, err := os.Open("aario.png")
    if err != nil {
        log.Fatalf("os.Open failed: %v", err)
    }
    img, _, err := image.Decode(f)
    if err != nil {
        log.Fatalf("image.Decode failed: %v", err)
    }
    //f.Close()              <---------------    forget to close it
}
```

启动后，观察句柄情况，大概有4个句柄
```shell
sh$ ls /proc/$pid/fd
|[
0 1 2 3 4
]|
```

使用并发测试，然后观察句柄情况，会发现非常多句柄产生，之后停止并发测试，会发现由于忘记关闭句柄，那么这些句柄会一直被占用，而如果句柄数超过 `ulimit -n` 那么就会导致进程死掉。
```shell
sh$ ls /proc/$pid/fd
|[
0 3 4   ...  120 402 ...
]|
```

而如果使用 `Close()` 关闭了句柄，使用并发测试依旧会发现生成了非常多句柄，而如果句柄数超过 `ulimit -n` 那么就会导致进程死掉。但是，当停止并发测试，会发现新打开的文件句柄慢慢全部被回收。

对于正常高并发导致句柄超了，简单的方案，可以考虑用监控子进程、supervisor 监控主进程等方式保活。